blocks/dump.py

```diff
@@ -0,0 +1,220 @@
+"""Dumping of the main loop object.
+
+The default way to save the training progress in Blocks is full
+serialization of the main loop object. This way is however often obstructed
+by various technical difficulties. Even a minor change in any of the
+libraries used might make it impossible to deserialize your experiment
+back.
+
+For this reason Blocks supports a cheaper but less reliable alternative
+called _dumping_. A dump of the main loop contains the most essential data
+from the training process: the model parameters and the log. In addition,
+to make training resumption possible, the iteration state is saved, that is
+the data stream and the epoch iterator.
+
+While current dumping mechanism still uses serialization, this is subject
+to be gradually changed, aiming to use only stable and simple data formats
+for dumps, such as for instance .npz files.
+
+"""
+import logging
+import os
+import os.path
+from collections import OrderedDict
+
+import dill
+import numpy
+
+from blocks.bricks.base import Brick
+from blocks.select import Selector
+
+logger = logging.getLogger(__name__)
+
+
+def save_parameter_values(param_values, path):
+    """Compactly save parameter values.
+
+    This is a thin wrapper over `numpy.savez`. It deals with
+    `numpy`'s vulnerability to slashes in file names.
+
+    Parameters
+    ----------
+    param_values : dict of (parameter name, numpy array)
+        The parameter values.
+    path : str of file
+        The destination for saving.
+
+    """
+    param_values = {name.replace("/", "-"): param
+                    for name, param in param_values.items()}
+    numpy.savez(path, **param_values)
+
+
+def load_parameter_values(path):
+    """Load parameter values saved by :fun:`save_parameters`.
+
+    This is a thin wrapper over `numpy.load`. It deals with
+    `numpy`'s vulnerability to slashes in file names.
+
+    Parameters
+    ----------
+    path : str or file
+        The source for loading from.
+
+    Returns
+    -------
+    A dictionary of (parameter name, numpy array) pairs.
+
+    """
+    source = numpy.load(path)
+    param_values = {name.replace("-", "/"): value
+                    for name, value in source.items()}
+    source.close()
+    return param_values
+
+
+def extract_parameter_values(bricks):
+    """Extract parameter values from a bricks hierarchy.
+
+    Parameters
+    ----------
+    bricks : Brick or Selector
+        The top bricks.
+
+    Returns
+    -------
+    A dictionary of (parameter name, numpy array) pairs.
+
+    """
+    if isinstance(bricks, Brick):
+        bricks = Selector([bricks])
+    if not isinstance(bricks, Selector):
+        raise ValueError
+    return OrderedDict([(name, variable.get_value(borrow=True))
```

Comment:
One way of doing it would be by a similar data structure as I use for the parameters. Basically, this would give brick a `parents` attribute (because you can have multiple) that is kept entirely in sync with the `children` attributes (so when you modify the children of a brick, they automatically get parents added/removed). That would then allow us to reconstruct the path from any brick.

However, this path is not unique as long as we allow one brick to be the child of multiple parents. How is this currently handled? I guess that you can just ignore it; you will simply save the identical parameters twice, and then reload them twice. But if want to create a unique ID, you need to make a decision as to what how you deal with multiple parents (e.g. choose the first parent, use both).